third_party/move/move-compiler-v2/src/pipeline/split_critical_edges_processor.rs

```diff
@@ -0,0 +1,193 @@
+        if fun_env.is_native() {
+            return data;
+        }
+        let mut transformer = SplitCriticalEdgesTransformation::new(data);
+        transformer.transform();
+        transformer.data
+    }
+
+    fn name(&self) -> String {
+        "SplitCriticalEdgesProcessor".to_owned()
+    }
+}
+
+struct SplitCriticalEdgesTransformation {
+    data: FunctionData,
+    // labels used in the original code and in the generated code
+    labels: BTreeSet<Label>,
+    srcs_count: BTreeMap<Label, usize>,
+}
+
+impl SplitCriticalEdgesTransformation {
+    pub fn new(data: FunctionData) -> Self {
+        let labels = Bytecode::labels(&data.code);
+        let srcs_count = count_srcs(&data.code);
+        Self {
+            data,
+            labels,
+            srcs_count,
+        }
+    }
+
+    /// Runs the transformation, which breaks critical edges with empty blocks.
+    fn transform(&mut self) {
+        let bytecodes = std::mem::take(&mut self.data.code);
+        for bytecode in bytecodes {
+            self.transform_bytecode(bytecode)
+        }
+    }
+
+    /// Transforms a bytecode
+    fn transform_bytecode(&mut self, bytecode: Bytecode) {
+        match bytecode {
+            Bytecode::Branch(attr_id, l0, l1, t) => self.transform_branch(attr_id, l0, l1, t),
+            // Edge of a `Jump` can never be critical since the source node only have one out edge
+            _ => self.emit(bytecode),
+        }
+    }
+
+    /// Transforms a branch instruction by splitting the critical out edges
+    pub fn transform_branch(&mut self, attr_id: AttrId, l0: Label, l1: Label, t: TempIndex) {
+        match (
+            self.split_critical_edge(attr_id, l0),
+            self.split_critical_edge(attr_id, l1),
+        ) {
+            (None, None) => self.emit(Bytecode::Branch(attr_id, l0, l1, t)),
+            (None, Some((l1_new, code))) => {
+                self.emit(Bytecode::Branch(attr_id, l0, l1_new, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code)), None) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1, t));
+                self.emit_codes(code)
+            },
+            (Some((l0_new, code0)), Some((l1_new, code1))) => {
+                self.emit(Bytecode::Branch(attr_id, l0_new, l1_new, t));
+                self.emit_codes(code0);
+                self.emit_codes(code1);
+            },
+        }
+    }
+
+    /// `label` is the target of some branch instruction
+    /// If label has multiple sources, returns
+    /// - a fresh label
+    /// - a new empty block that
+    ///     - starts with the fresh label
+    ///     - jumps to `label`
+    /// otherwise returns `None`
+    fn split_critical_edge(
+        &mut self,
+        attr_id: AttrId,
+        label: Label,
+    ) -> Option<(Label, Vec<Bytecode>)> {
+        // label is in `srcs_count` by construction
+        if *self.srcs_count.get(&label).expect("srcs count") > 1 {
```

Comment:
Would be helpful to have some notes added arguing the correctness: we are transforming code as we go on, but we depend on the `srcs_count` of the old code. I think this is okay because this transformation does not change the `srcs_count` (it replaces each incoming edge with another, or it does not do anything).

Reply:
done